tianshou/data/batch.py

```diff
@@ -502,49 +546,92 @@ def to_torch(self, dtype: Optional[torch.dtype] = None,
                 v.to_torch(dtype, device)
 
     def cat_(self,
-             batches: Union['Batch', List[Union[dict, 'Batch']]]) -> None:
+             batches: Union['Batch', List[Union[dict, 'Batch']]],
+             lens: Optional[List[int]] = None
```

Comment:
I don't like such the idea of using hint.

Reply:
A better solution is welcome. But it seems unavoidable for the case I give:

```python
x=Batch(a=Batch(a=np.random.randn(3, 4)), b=np.random.randn(3, 4))
y = Batch(a=Batch(a=Batch()), b=np.random.randn(3, 4))
Batch.cat([x, y])
```

Reply:
Yes, this issue was already discussed in the past. It is possible. You need to do something like this:

`
def stack(batches):
    def stack_impl(batches, lens):
        [...] # actual stack implementation

    lens = [...] # compute len here
    return stack_impl(batches, lens)
```

This way, it hides the implementation of stack that is used internally, because the `lens` parameter should not appears at top level in my opinion.

Reply:
Agree that ``lens `` should not be exposed to users.

Reply:
B y the way, take advantage of this structure to make `lens` argument required and not optional in `cat_impl`, and move the computation of the initial value outside `cat_impl`. It will make the complete implementation clearer and simpler.

Reply:
> But creating a nested function would make the indentation too deep. I changed it into an internal function __cat.

As you wish. It is just an implementation detail.